Engine/ac/string.cpp

```diff
@@ -427,30 +427,55 @@ int StringToInt(const char*stino) {
     return atoi(stino);
 }
 
-int StrContains (const char *s1, const char *s2) {
-    VALIDATE_STRING (s1);
-    VALIDATE_STRING (s2);
-    char *tempbuf1 = ags_strdup(s1);
-    char *tempbuf2 = ags_strdup(s2);
-    ustrlwr(tempbuf1);
-    ustrlwr(tempbuf2);
+int String_IndexOf (const char *thisString, const char *lookForText, bool caseSensitive, int startIndex, int endIndex) {
+    VALIDATE_STRING(thisString);
+    VALIDATE_STRING(lookForText);
 
-    char *offs = ustrstr(tempbuf1, tempbuf2);
+    const auto &header = ScriptString::GetHeader(thisString);
+    if ((startIndex < 0) || ((size_t) startIndex > header.ULength)) {
+        debug_script_warn("String.IndexOf: invalid start index %d.", startIndex);
+        return -1;
+    }
+    if (endIndex < startIndex) {
+        debug_script_warn("String.IndexOf: end index (%d) is smaller than start index (%d).", endIndex, startIndex);
+        return -1;
+    }
+    if (endIndex == -1 || endIndex >= header.ULength)
+        endIndex = header.ULength;
 
-    if (offs == nullptr)
+    size_t start = uoffset(thisString, startIndex);
+    size_t end = uoffset(thisString, endIndex);
+    size_t byte_len = end - start;
+    size_t u_len = endIndex - startIndex;
+
+    auto buf = ScriptString::CreateBuffer(byte_len, u_len);
```

Comment:
There's virtually no reason to use ScriptString::CreateBuffer here, this function is only ever needed when you need to create and return a new ScriptString object. This function creates a buffer with a meta header prepended to it.
Here you just run a standard string algorithm over a pair of buffers, so ags_strdup is correct.

Reply:
Alright, I see, you need a substring.
In such case you could simply use a common String, which has a constructor accepting char pointer and length.
You may use String for both buffers, and it will automatically delete itself too.

Reply:
> then it may be convenient to have helper functions mimicing "strdup" but return a C++ container

I think this sounds best but wasn't sure where to put this helper. I thought about mimicking but taking (optional?) start and end parameters so it can either duplicate or substring.

Reply:
> I think this sounds best but wasn't sure where to put this helper

StrUtil namespace is a good place for this, in string_utils.h/cpp

Reply:
> If using String, it will need to be added unicode-aware MakeLower/Upper methods (similar to existing String::ReverseUTF8).

Erh, just to be sure I am going the right way, the reason I thought to not go this way is because string.cpp is currently independent from allegro, so the only way to keep it this way would be to duplicate the utolower/utoupper functions that apply on a single char.

Reply:
> the reason I thought to not go this way is because string.cpp is currently independent from allegro, so the only way to keep it this way would be to duplicate the utolower/utoupper functions that apply on a single char.

Hmm, right, getting allegro dependency there will be not good.
So either the raw buffer way where a substring is returned as `std::vector<char>`, or if in String then it has to be fully recreated there, using at most only Utf8 utilities header.

Reply:
I ended up adding to the StrUtil namespace.